### ISSUE TITLE
Problem with tables that have an ID column

### ISSUE THREAD

**Issue by edsonlima** [composite-primary-keys/composite_primary_keys#319]
Hi,
Recently I tried to update to version 8.1.1 because I faced #291.
But, unfortunately, a lot of my tests broke because I have ID columns in some tables.

For example, considering the following models:

``` ruby
class Person < ActiveRecord::Base
  self.primary_keys = :id, :company_id

  has_many :documents, class_name: 'PersonDocument', primary_key: [:id, :company_id],
                   foreign_key: [:person_id, :company_id]

  has_many :wtf_documents, ->(record) { where(person_id: record.id[0][0],
                                         company_id: record.company_id) },
           class_name: 'PersonDocument',
           foreign_key: ['person_id', 'company_id']

end

class PersonDocument < ActiveRecord::Base
  self.primary_keys = :id, :company_id
end
```

With tables:

``` ruby
class CreatePeople < ActiveRecord::Migration
  def change
    create_table :people do |t|
      t.string :name
      t.references :company, index: true
      t.timestamps null: false
    end
  end
end

class CreatePersonDocuments < ActiveRecord::Migration
  def change
    create_table :person_documents do |t|
      t.string :document_number
      t.integer :person_id, index: true
      t.integer :company_id, index: true
      t.timestamps null: false
    end
  end
end
```

After creating some records:

``` ruby
Person.create(company_id: 1, name: 'Person name')
PersonDocument.create(person_id: 1, company_id:1, document_number: '4as65a4s654')
```

If I try to get the object id, this is my result:

``` ruby
pe = Person.first
#SELECT  "people".* FROM "people"  ORDER BY "people"."company_id" ASC LIMIT 1
# => #<Person id: [1, 1], name: "Person name", company_id: 1, 
# created_at: "2015-08-20 00:21:33", updated_at: "2015-08-20 00:21:33">
pe.id
# => [[1, 1], 1] 
```

So, if I try to get the has_many association this is my result:

``` ruby
pe.documents
# SELECT "person_documents".* FROM "person_documents" WHERE 
# ("person_documents"."person_id" = NULL AND "person_documents"."company_id" = 1)
# => #<ActiveRecord::Associations::CollectionProxy []>
```

Even though I try to pass the ids of  has_many association directly, the problem persist:

``` ruby
pe.wtf_documents
# SELECT "person_documents".* FROM "person_documents" WHERE 
#("person_documents"."person_id" = NULL AND "person_documents"."company_id" = 1) AND 
# "person_documents"."person_id" = $1 AND "person_documents"."company_id" = $2  
#[["person_id", 1], ["company_id", 1]]
# => #<ActiveRecord::Associations::CollectionProxy []>
```

But, as you can see in this case, duplicates the 'where' clause.

On the other hand, none of these problems occurs if table no have an ID column. In this case, the object id is returned as expected, and associations are loaded correctly:

``` ruby
wi = WithoutIdPerson.first
# SELECT  "without_id_people".* FROM "without_id_people"  ORDER BY 
#"without_id_people"."company_id" ASC LIMIT 1
# => #<WithoutIdPerson my_pk: 1, name: "No ID Person", company_id: 1,
# created_at: "2015-08-20 00:47:12", updated_at: "2015-08-20 00:47:12"> 
wi.id
# => [1, 1] 
```

For now, I went back to version 8.0.0 with Rails 4.2.0 and did a "workaround" to #291... 
Any help is very appreciated, thanks.

**Comment by edsonlima** [comment]
Just for clarification, when I call object.id, the last value in array is the ID of second key field:
```ruby
pe = Person.first
# SELECT  "people".* FROM "people"  ORDER BY "people"."company_id" ASC LIMIT 1
# => #<Person id: [1, 1], name: "Person name", company_id: 1, 
# created_at: "2015-08-20 00:21:33", updated_at: "2015-08-20 00:21:33"> 
pe.company_id = 3
# => 3 
pe.save
#   (0.5ms)  BEGIN
 # SQL (0.6ms)  UPDATE "people" SET "company_id" = $1, "updated_at" = $2 WHERE 
# ("people"."id" = NULL AND "people"."company_id" = 3)  [["company_id", 3], ["updated_at", 
# "2015-08-20 13:46:55.911723"]]
#  (0.5ms)  COMMIT
# => true 
pe.id
# => [[1, 3], 3] 
```

**Comment by skatscher** [comment]
Dear edsonlima,
did you find a solution to your problem? We came across a similar problem in our project.
Thank you!

**Comment by edsonlima** [comment]
Hi @skatscher
Unfortunately, no.
For now, we went back to 8.0.0 with rails 4.2.0, with #291 :(
To workaround #291, in some cases we added a method like this:

```ruby
class PersonDocument < ActiveRecord::Base
  self.primary_keys = :id, :company_id
  belongs_to :person primary_key: [:person_id, :company_id]
                   foreign_key: [:id, :company_id]

  attr_reader :person

  def person
    @person ||= Person.find([:person_id, :company_id])
  end

end
``` 

Honestly, every time I see this code, I wonder how much it increases our WTF/per minute (http://www.osnews.com/story/19266/WTFs_m) ;)

**Comment by tbauscher** [comment]
I'm experiencing this same problem.

**Comment by hayesr** [comment]
I just overwrote the id method that CPK provides, and that seems to be working for me. In my model:

```Ruby
def id
  attr_names = self.class.primary_keys
  attr_map = attr_names.map do |attr_name|
    if attr_name == 'id'
      attributes['id']
    else
      read_attribute(attr_name)
    end
  end
  ::CompositePrimaryKeys::CompositeKeys.new(attr_map)
end
```

Using `8.1.1`

**Comment by lichtamberg** [comment]
Is there a fix for this problem? Started using this gem and within minutes my first issue lookup :D

**Comment by cfis** [comment]
Fixed by #370.
